### ISSUE TITLE
Inheritance support

### ISSUE THREAD

**Issue by tfactor2** [victools/jsonschema-generator#407]
Hi,
I'm using the library to generate schemas from classes with hierarchy.
As far as we understand,  definitions for parent properties are always duplicated, right? If yes, do you have any plans to support the reference to the parent class (extension)?

**Current behavior:**
For these classes:
```java
@Data
public class Inheritance {

    @NotNull
    private String schemaVersion = "1.0";
    private Inheritance.Parent parent;
    private Inheritance.Child1 child1;

    @Data
    public static class Parent {
        private String parentField;
    }

    @Data
    @EqualsAndHashCode(callSuper = true)
    public static class Child1 extends Parent {

        private String child1NestedField;
    }
}
```
The library will generate the following schema:
```json
{
{
  "$schema" : "https://json-schema.org/draft/2020-12/schema",
  "$ref" : "#/$defs/Inheritance",
  "$defs" : {
    "Child1" : {
      "type" : "object",
      "properties" : {
        "parentField" : {
          "type" : "string"
        },
        "child1NestedField" : {
          "type" : "string"
        }
      }
    },
    "Inheritance" : {
      "type" : "object",
      "properties" : {
        "schemaVersion" : {
          "type" : "string",
          "default" : "1.1"
        },
        "parent" : {
          "$ref" : "#/$defs/Parent"
        },
        "child1" : {
          "$ref" : "#/$defs/Child1"
        }
      },
      "required" : [ "schemaVersion" ]
    },
    "Parent" : {
      "type" : "object",
      "properties" : {
        "parentField" : {
          "type" : "string"
        }
      }
    }
  }
}
```

Notice that the `parentField` is duplicated.

**It would be nice to get something like this:**

```json
"$defs" : {
    "Child1" : {
      "type" : "object",
      "$ref" : "#/$defs/Parent",
      "properties" : {
        "child1NestedField" : {
          "type" : "string"
        }
      }
    },
```

**Comment by CarstenWickner** [comment]
Hi @tfactor2,

Unfortunately, that is currently not supported.
I didn't find a satisfying solution for children that override parent properties when considering something along these lines some time ago. Certain JSON schema features are not compatible with this kind of structure (e.g., `additionalProperties` only considers the single schema in which it is defined, not any referenced subschemas).
Also, I assumed the consumer of a JSON would receive a flat object and not care about deeply nested Java classes representing it. Personally, I find such a schema harder to read.

That said, I get where you're coming from and am open to a PR if you want to take a stab at this feature.
So far, it didn't seem worth the trouble to me, i.e., I wasn't planning on introducing this myself.

I'm fairly confident though, that you could achieve this through configuration alone, but it would need to circumvent certain aspects of the built-in generation process (namely ignoring all properties defined in a parent) and then re-implementing that (by referencing the parent explicitly).

**Comment by tfactor2** [comment]
Hi @CarstenWickner ,

Thanks for the quick response.

`additionalProperties` according to [docs](https://swagger.io/docs/specification/data-models/dictionaries/) represent dictionaries, it is natural that they can not be inherited, IMHO.

The motivation behind the parent reference for us lies in the email template case:
We have a dozen of email templates with typed placeholders inherited from the common one containing the base data (the base template itself has multiple fields).
Thus, duplicating parent fields in every child template will not only make the schema much bigger, but also will at some extent lead to the lost in semantics.

----

My idea in the library modification direction was to make the engine differentiate between own and inherited fields first, and then optionally put a reference to parent.
But I didn't find a possibility to do it via `classmate` that is used for introspection in the library.

**Comment by CarstenWickner** [comment]
Hi @tfactor2,
I see where you're coming from and understand the wish to avoid excessive duplication.

As for identifying own vs. inherited properties, you should be able to tell from a given field/method what the declaring type is, which may or may not match the specific type a (sub)schema is being generated for. When targeting a specific use case, such as a base email template, that configuration should be straightforward (much more so than any generic solution).

As per my previous comment: the separation/reference to the parent should be possible as well through configuration alone.
Let me know, if you require assistance there.

**Comment by tfactor2** [comment]
We decided to flatten schemas
.
As we are building the library that itself uses the `jsonschema-generator`, providing schema-specific customizations is not an option for us.
Mail template is just one of the cases, but we need a generic solution.


On the other hand, I understood your point about the complexity of the task, so I believe the issue could be closed as Won't Fix.

Thanks for your assistance.

**Comment by CarstenWickner** [comment]
Identifying properties declared in supertypes still sounds like a good idea, to make this kind of configuration easier in the future.
While you may be fine with the flattened schema (for now), others might still have the same requirement.

I'll leave this open and potentially look into it.
I'd probably not offer full built-in support, but your use case is still valid and should be straightforward (through configuration) if the necessary information is present.
Perhaps a simple example to show case it will be sufficient then.

Thanks for bringing it up.

**Comment by CarstenWickner** [comment]
Hi @tfactor2,

I've made some changes to allow identifying members that are declared directly or in a super type.
The [PR #415](https://github.com/victools/jsonschema-generator/pull/415) also includes an example of what you originally requested.
As it comes with the risk of running into an infinite loop, I'd say your decision to flatten the schemas was a good one.

**Comment by tfactor2** [comment]
Great, thank you!

**Comment by CarstenWickner** [comment]
The changes mentioned above have been released as part of [v4.33.0](https://github.com/victools/jsonschema-generator/releases/tag/v4.33.0) now.
